### PR TITLE
insert/remove range bounds

### PR DESCRIPTION
**Pull request by oliverdding**
Implimentation of `insert_range` and `remove_range` functions that accept `RangeBounds` instead of `Range`.

This PR may fix [#5] and would fix [#113].

Done:

* [x] Switch inner parameters to `RangeBounds`
   * [x] `insert_range`
   * [x] `remove_range`
 * [x] Fix boundary condition in `insert_range`/`remove_range`

TODO:

* [x] Add more test cases
* [x] switching to `RangeBounds` of container
   * [x] `insert_range`
   * [x] `remove_range`

### CODE REVIEW

File: src/bitmap/container.rs [hunk 1]
```diff
@@ -38,13 +39,10 @@ impl Container {
         }
     }
 
-    pub fn insert_range(&mut self, range: Range<u16>) -> u64 {
-        // If the range is larger than the array limit, skip populating the
-        // array to then have to convert it to a bitmap anyway.
-        if matches!(self.store, Store::Array(_)) && range.end - range.start > ARRAY_LIMIT as u16 {
-            self.store = self.store.to_bitmap()
-        }
```
Comment (Kerollmops) [inline src/bitmap/container.rs:46]: Did you forget to bring this part back in your PR?

File: src/bitmap/container.rs [hunk 2]
```diff
@@ -68,12 +66,11 @@ impl Container {
         }
     }
 
-    pub fn remove_range(&mut self, start: u32, end: u32) -> u64 {
-        debug_assert!(start <= end);
-        if start == end {
-            return 0;
-        }
```
Comment (Kerollmops) [inline src/bitmap/container.rs:75]: As we are in the `Container` we should accept a `RangeInclusive<u16>` here and return 0 if it `is_empty`.

File: src/bitmap/container.rs [hunk 3]
```diff
@@ -38,13 +39,10 @@ impl Container {
...
-        // If the range is larger than the array limit, skip populating the
-        // array to then have to convert it to a bitmap anyway.
-        if matches!(self.store, Store::Array(_)) && range.end - range.start > ARRAY_LIMIT as u16 {
-            self.store = self.store.to_bitmap()
-        }
-
+    pub fn insert_range<R>(&mut self, range: R) -> u64
+    where
+        R: RangeBounds<u16>,
+    {
```
Comment (Kerollmops) [inline src/bitmap/container.rs:45]: As we are in the `Container` we should accept a `RangeInclusive` here.

File: src/bitmap/inherent.rs [hunk 4]
```diff
@@ -254,61 +254,62 @@ impl RoaringBitmap {
...
-                    if self.containers[index].len == 0 {
-                        self.containers.remove(index);
-                        continue;
-                    }
-                }
-            }
-            index += 1;
-        }
-        result
+        todo!()
```
Comment (Kerollmops) [inline src/bitmap/inherent.rs:257]: Did you forget to bring back this part in your PR?

File: src/bitmap/store.rs [hunk 5]
```diff
@@ -43,40 +45,72 @@ impl Store {
         }
     }
 
-    pub fn insert_range(&mut self, range: Range<u16>) -> u64 {
-        // A Range is defined as being of size 0 if start >= end.
-        if range.is_empty() {
+    pub fn insert_range<R>(&mut self, range: R) -> u64
+    where
+        R: RangeBounds<u16>,
+    {
```
Comment (Kerollmops) [inline src/bitmap/store.rs:51]: As we are in the `Store` we should accept a `RangeInclusive<u16>` here, that would simplify the usage of the library. Because we know that we are in a finished range of numbers i.e. `0..=u32::MAX`.

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Thank you, @oliverdding for this PR.

Is it finished, yet? I am not sure about that as there is a `todo!` in the middle 😄

Would it be possible to see [`RangeInclusive<u16>`](https://doc.rust-lang.org/std/ops/struct.RangeInclusive.html) ranges for the `Container` and `Store` structs, converting the [`RangeBounds<u32>`](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) traits into `RangeInclusive<u16>` before and checking that the conversion [doesn't creates an empty range](https://doc.rust-lang.org/std/ops/struct.RangeInclusive.html#method.is_empty).

File: src/bitmap/container.rs (hunk 1 above)
Comment (oliverdding) [reply src/bitmap/container.rs:46]: I thought that I only need to handle the inner one 😂, so I didn't handle the container. I would add another commit later.

File: src/bitmap/container.rs (hunk 3 above)
Comment (oliverdding) [reply src/bitmap/container.rs:45]: Infact, I think it's also need to use `RangeBounds` here since it's `pub` one. This may break the users who directly use this api. The same to the store.

**Comment by oliverdding** [comment]
Thanks for reviews @Kerollmops 

I'm still working on it, and I created this PR because I need some advice on my direction in order to avoid too many changes.

I've seen the comments you made, and I got a fatal problem:

Should `insert_range`/`remove_range` of store, container use the `RangeBounds` or `RangeInclusive`?

File: src/bitmap/store.rs [hunk 6]
```diff
@@ -146,37 +186,63 @@ impl Store {
...
-                let start_key = key(start as u16) as usize;
-                let start_bit = bit(start as u16) as u32;
-                // end_key is inclusive
-                let end_key = key((end - 1) as u16) as usize;
-                let end_bit = bit(end as u16) as u32;
+                let (start_key, start_bit) = (key(start), bit(start));
+                let (end_key, end_bit) = (key(end), bit(end));
 
                 if start_key == end_key {
-                    let mask = (!0u64 << start_bit) & (!0u64).wrapping_shr(64 - end_bit);
```
Comment (oliverdding) [inline src/bitmap/store.rs:170]: Oops, I forget to revert this. In fact, `!0u64` is equal to `u64::MAX`.

**Comment by Kerollmops** [comment]
> Should `insert_range`/`remove_range` of store, container use the `RangeBounds` or `RangeInclusive`?

Everything that is non-public (`Store/Container::insert_range/remove_range`) should use the `RangeInclusive` struct to ease the internal usage of them and everything that is public (`RoaringBitmap/RoaringTreemap::insert_range/remove_range`) should use the more user-friendly `RangeBounds<u32/u64>` trait. The public methods should convert the trait into the `RangeInclusive<u16>` structs to call the non-public methods.

**Comment by oliverdding** [comment]
Hi @Kerollmops 

Sorry for a mess here, now I think I have totally got your point, and should I close this PR and re-arange my code again or just continue to modify?

**Comment by Kerollmops** [comment]
> I think I have totally got your point now 😊. But it's a mess here, I would like to closed this PR and re-arange the code again.

Don't do that, haha, you can simply rewrite the commit and push force (`git push --force`) on this branch, no worries. Just ping me when you are done :)

File: src/bitmap/container.rs (hunk 3 above)
Comment (Kerollmops) [reply src/bitmap/container.rs]: It is not public, as I told you in https://github.com/RoaringBitmap/roaring-rs/issues/113#issuecomment-925609379. Can we please continue speaking in only one place, it would be easier for me to follow the discussion, I would like that we exchange in this PR ;)

**Comment by Kerollmops** [comment]
BTW it seems that you miss the point where you started you branch, you should either:
 - Create your branch and commit from the end of this PR's branch #92 and make [sure this PR base branch is `insert-remove-range-bounds`](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request). (preferred solution)
 - Create your branch and commit from the `master` branch.

**Comment by oliverdding** [comment]
Hi @Kerollmops 

After some hard debuging time, I have finished all the jobs. But this PR is mess up by my hard resetting and forcing pushing...

I'm sorry, this is my first PR, and I'm not clear what happens 😭

In a nutshell, can I restart a PR?

**Comment by Kerollmops** [comment]
> I'm sorry, this is my first PR, and I'm not clear what happens 😭
> 
> In a nutshell, can I restart a PR?

Haha, no worries, go on, create a new PR but make sure you correctly set the base branch 😄
